In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [49]:
df = pd.read_csv('/content/drive/MyDrive/lab3/car_info.tsv',sep='\t')


In [4]:
import requests
import numpy as np
import csv

In [5]:
len(df)

5073

In [6]:
df.head(10)

mark  price  produced_year  imported_year  \
0            Subaru Forester   40.0           2017           2019   
1            Toyota Prius 20    8.0           2004           2015   
2             Toyota Alphard   43.0           2013           2022   
3             Toyota Harrier   22.0           2009           2013   
4  Toyota Land Cruiser Prado  101.0           2016           2024   
5               Nissan NV200   15.0           2013           2024   
6        Subaru XV Crosstrek   35.0           2018           2024   
7                   Lexus IS   40.5           2014           2024   
8            Toyota Prius 20   20.3           2007           2023   
9                   Lexus IS   40.0           2014           2020   

      distance motor_volume   color           type   hurd hodolguur  
0    93000 км.        2.5 л     Хар           Жийп  Буруу    Бензин  
1  1111111 км.        0.5 л   Бусад  Суудлын тэрэг  Буруу   Хайбрид  
2   154279 км.        2.4 л     Хар     Гэр бүлийн  Буруу    Бензин  
3   150000 км.        3.5 л  Шаргал           Жийп  Буруу    Бензин  
4    54000 км.        2.7 л  Саарал           Жийп  Буруу    Бензин  
5      250 км.        1.6 л  Саарал  Суудлын тэрэг  Буруу    Бензин  
6      150 км.        2.0 л  Шаргал     Гэр бүлийн  Буруу    Бензин  
7   160000 км.        2.5 л     Хар  Суудлын тэрэг  Буруу   Хайбрид  
8    74000 км.        1.5 л  Шаргал  Суудлын тэрэг  Буруу   Хайбрид  
9      147 км.        2.0 л     Хар  Суудлын тэрэг  Буруу   Хайбрид

In [50]:
df['distance'] = df['distance'].str.replace(' км.', '').astype(float)

<ipython-input-50-3c5e2011e05f>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  df['distance'] = df['distance'].str.replace(' км.', '').astype(float)


In [51]:
df['motor_vol'] = df['motor_volume'].str.extract(r'(\d+\.\d+|\d+)').astype(float)
df.drop('motor_volume', axis=1, inplace=True)

In [52]:
df.head()

mark  price  produced_year  imported_year   distance  \
0            Subaru Forester   40.0           2017           2019    93000.0   
1            Toyota Prius 20    8.0           2004           2015  1111111.0   
2             Toyota Alphard   43.0           2013           2022   154279.0   
3             Toyota Harrier   22.0           2009           2013   150000.0   
4  Toyota Land Cruiser Prado  101.0           2016           2024    54000.0   

    color           type   hurd hodolguur  motor_vol  
0     Хар           Жийп  Буруу    Бензин        2.5  
1   Бусад  Суудлын тэрэг  Буруу   Хайбрид        0.5  
2     Хар     Гэр бүлийн  Буруу    Бензин        2.4  
3  Шаргал           Жийп  Буруу    Бензин        3.5  
4  Саарал           Жийп  Буруу    Бензин        2.7

color баганыг хасав. Учир нь машины үнэд өнгө нөлөөлөхгүй гэж үзлээ.

In [53]:
df.drop('color', axis=1, inplace=True)

distance Баганы утгууд зарим нь худлаа бичигдсэн байна. Жишээ нь 2013 онд орж ирсэн машин 250 км явсан гэсэн байна. Энэ мэт датануудыг 1000 аар үржлээ.

In [54]:
def multiply_if_less_than_1000(value):
    if value < 1000:
        return value * 1000
    else:
        return value

df['distance'] = df['distance'].apply(multiply_if_less_than_1000)

In [55]:
df.head()

mark  price  produced_year  imported_year   distance  \
0            Subaru Forester   40.0           2017           2019    93000.0   
1            Toyota Prius 20    8.0           2004           2015  1111111.0   
2             Toyota Alphard   43.0           2013           2022   154279.0   
3             Toyota Harrier   22.0           2009           2013   150000.0   
4  Toyota Land Cruiser Prado  101.0           2016           2024    54000.0   

            type   hurd hodolguur  motor_vol  
0           Жийп  Буруу    Бензин        2.5  
1  Суудлын тэрэг  Буруу   Хайбрид        0.5  
2     Гэр бүлийн  Буруу    Бензин        2.4  
3           Жийп  Буруу    Бензин        3.5  
4           Жийп  Буруу    Бензин        2.7

In [56]:
df['distance'] = df['distance'].replace(0, 50000)

In [57]:
avg_motor_vol = df['motor_vol'].mean()
avg_motor_vol

2.3868604419669524

In [58]:
df['motor_vol'] = df['motor_vol'].replace(0, avg_motor_vol)

In [59]:
df['distance_num'] = df['distance'].astype(float)
df.drop('distance', axis=1, inplace=True)

In [60]:
df.head()

mark  price  produced_year  imported_year  \
0            Subaru Forester   40.0           2017           2019   
1            Toyota Prius 20    8.0           2004           2015   
2             Toyota Alphard   43.0           2013           2022   
3             Toyota Harrier   22.0           2009           2013   
4  Toyota Land Cruiser Prado  101.0           2016           2024   

            type   hurd hodolguur  motor_vol  distance_num  
0           Жийп  Буруу    Бензин        2.5       93000.0  
1  Суудлын тэрэг  Буруу   Хайбрид        0.5     1111111.0  
2     Гэр бүлийн  Буруу    Бензин        2.4      154279.0  
3           Жийп  Буруу    Бензин        3.5      150000.0  
4           Жийп  Буруу    Бензин        2.7       54000.0

In [61]:

def preprocess(app_set):
    ret = pd.DataFrame(columns = ["mark", "price", "produced_year", "imported_year", "type", "hurd", "hodolguur", "distance_num", "motor_vol"])
    for index, row in app_set.iterrows():
        if row['hurd'] == 'Буруу':
          leas = False
        else:
          leas = True
        if row['distance_num'] > 1000000:
          road = row['distance_num'] / 10
        else:
          road = row['distance_num']

        ret = ret.append({'mark': row['mark'], 'price': row['price'], 'produced_year': row['produced_year'],
                          'imported_year': row['imported_year'], 'type': row['type'], 'hurd': leas,
                          'hodolguur': row['hodolguur'], 'distance_num': road, 'motor_vol' : row['motor_vol']}, ignore_index = True)
    return ret


In [ ]:
finally_data = preprocess(df)

In [63]:
finally_data.head()

mark  price produced_year imported_year  \
0            Subaru Forester   40.0          2017          2019   
1            Toyota Prius 20    8.0          2004          2015   
2             Toyota Alphard   43.0          2013          2022   
3             Toyota Harrier   22.0          2009          2013   
4  Toyota Land Cruiser Prado  101.0          2016          2024   

            type   hurd hodolguur  distance_num  motor_vol  
0           Жийп  False    Бензин       93000.0        2.5  
1  Суудлын тэрэг  False   Хайбрид      111111.1        0.5  
2     Гэр бүлийн  False    Бензин      154279.0        2.4  
3           Жийп  False    Бензин      150000.0        3.5  
4           Жийп  False    Бензин       54000.0        2.7

In [70]:
finally_data.drop('type', axis=1, inplace=True)
finally_data.drop('hurd', axis=1, inplace=True)

In [74]:
finally_data.head()

mark  price produced_year imported_year hodolguur  \
0            Subaru Forester   40.0          2017          2019    Бензин   
1            Toyota Prius 20    8.0          2004          2015   Хайбрид   
2             Toyota Alphard   43.0          2013          2022    Бензин   
3             Toyota Harrier   22.0          2009          2013    Бензин   
4  Toyota Land Cruiser Prado  101.0          2016          2024    Бензин   

   distance_num  motor_vol  
0       93000.0        2.5  
1      111111.1        0.5  
2      154279.0        2.4  
3      150000.0        3.5  
4       54000.0        2.7

In [38]:
!pip install -U sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 163.3/163.3 kB 3.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 27.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 37.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.1/14.1 MB 50.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 731.7/731.7 MB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.6/410.6 MB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 121.6/121.6 MB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.5/56.5 MB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 124.2/124.2 MB 8.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 196.0/196.0 MB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 166.0/166.0 MB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.1/99.1 kB 11.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━

In [47]:
from huggingface_hub import notebook_login
notebook_login()

In [48]:
from sentence_transformers import SentenceTransformer
sen_model = SentenceTransformer('joeddav/xlm-roberta-large-xnli')

config.json:   0%|          | 0.00/734 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.24G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

In [134]:
sentences = []
y = []
for index, row in df.iterrows():
    sentences.append(row['mark'] + ', ' + str(row['produced_year']) + ' онд үйлдвэрлэсэн, ' + str(row['imported_year']) + ' онд орж ирсэн, ' + str(row['distance_num']) + ' км явсан, ' + str(row['motor_vol']) + ' л мотор')
    y.append(row['price'])


In [135]:
sentences[:10]

['Subaru Forester, 2017 онд үйлдвэрлэсэн, 2019 онд орж ирсэн, 93000.0 км явсан, 2.5 л мотор',
 'Toyota Prius 20, 2004 онд үйлдвэрлэсэн, 2015 онд орж ирсэн, 1111111.0 км явсан, 0.5 л мотор',
 'Toyota Alphard, 2013 онд үйлдвэрлэсэн, 2022 онд орж ирсэн, 154279.0 км явсан, 2.4 л мотор',
 'Toyota Harrier, 2009 онд үйлдвэрлэсэн, 2013 онд орж ирсэн, 150000.0 км явсан, 3.5 л мотор',
 'Toyota Land Cruiser Prado, 2016 онд үйлдвэрлэсэн, 2024 онд орж ирсэн, 54000.0 км явсан, 2.7 л мотор',
 'Nissan NV200, 2013 онд үйлдвэрлэсэн, 2024 онд орж ирсэн, 250000.0 км явсан, 1.6 л мотор',
 'Subaru XV Crosstrek, 2018 онд үйлдвэрлэсэн, 2024 онд орж ирсэн, 150000.0 км явсан, 2.0 л мотор',
 'Lexus IS, 2014 онд үйлдвэрлэсэн, 2024 онд орж ирсэн, 160000.0 км явсан, 2.5 л мотор',
 'Toyota Prius 20, 2007 онд үйлдвэрлэсэн, 2023 онд орж ирсэн, 74000.0 км явсан, 1.5 л мотор',
 'Lexus IS, 2014 онд үйлдвэрлэсэн, 2020 онд орж ирсэн, 147000.0 км явсан, 2.0 л мотор']

In [136]:
sentence_vectors = sen_model.encode(sentences)

In [122]:
from sklearn.linear_model import LinearRegression
model = LinearRegression(fit_intercept=True)

In [137]:
x_ = sentence_vectors

In [138]:
model.fit(x_, y)

LinearRegression()

In [159]:
import re
def extract_car_info(sentence):
    patterns = {
        'mark': r'(.*?)\,',
        'produced_year': r'(\d+)\s+онд\s+үйлдвэрлэсэн',
        'imported_year': r'(\d+)\s+онд\s+орж\s+ирсэн',
        'distance_num': r'(\d+)\s+км\s+явсан',
        'motor_vol': r'(\d+(\.\d+)?)\s+л\s+мотор'
    }

    car_info = {key: [] for key in patterns}

    for key, pattern in patterns.items():
        match = re.search(pattern, sentence)
        if match:
            car_info[key].append(match.group(1))
        else:
            car_info[key].append(None)

    return car_info

sentence = "Toyota Land Cruiser 300, 2019 онд үйлдвэрлэсэн, 2021 онд орж ирсэн, 54000 км явсан, 3.0 л мотор"
car_info = extract_car_info(sentence)
print(car_info)

{'mark': ['Toyota Land Cruiser 300'], 'produced_year': ['2019'], 'imported_year': ['2021'], 'distance_num': ['54000'], 'motor_vol': ['3.0']}


In [160]:
def predictEstimate(sentence):
  p_embeddings = sen_model.encode([sentence])
  d = [extract_car_info(sentence)]
  p_ = []
  for emb in p_embeddings:
    #B = [d[0]]
    p_.append(emb)
  y_prediction =  model.predict(p_)
  return y_prediction[0]

In [162]:
print(predictEstimate("Lexus IS, 2014 онд үйлдвэрлэсэн, 2024 онд орж ирсэн, 160000.0 км явсан, 2.5 л мотор"))

49.457886


In [163]:
print(predictEstimate("Toyota Prius 20, 2019 онд үйлдвэрлэсэн, 2020 онд орж ирсэн, 100000.0 км явсан, 1.8 л мотор"))

26.253418


In [158]:
print(predictEstimate("Toyota Land Cruiser 100, 2021 онд үйлдвэрлэсэн, 2023 онд орж ирсэн, 80000 км явсан, 4.5 л мотор"))

147.24182
